presto-main/src/test/java/com/facebook/presto/sql/TestSqlToRowExpressionTranslator.java

```diff
@@ -14,38 +14,105 @@
 import com.facebook.presto.sql.tree.CoalesceExpression;
 import com.facebook.presto.sql.tree.Expression;
 import com.facebook.presto.sql.tree.LongLiteral;
 import com.facebook.presto.sql.tree.NodeRef;
 import com.facebook.presto.type.TypeRegistry;
 import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
+import java.util.Map;
+
 import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
 import static com.facebook.presto.metadata.FunctionKind.SCALAR;
 import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.sql.planner.LiteralInterpreter.toExpression;
+import static com.facebook.presto.transaction.TransactionManager.createTestTransactionManager;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
 
 public class TestSqlToRowExpressionTranslator
 {
+    private Metadata metadata;
+
+    @BeforeClass
+    public void setUp()
+    {
+        metadata = new MetadataManager(
+                new FeaturesConfig(),
+                new TypeRegistry(),
+                new BlockEncodingManager(new TypeRegistry()),
+                new SessionPropertyManager(),
+                new SchemaPropertyManager(),
+                new TablePropertyManager(),
+                createTestTransactionManager(new CatalogManager()));
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        metadata = null;
+    }
+
     @Test(timeOut = 10_000)
     public void testPossibleExponentialOptimizationTime()
     {
-        TypeManager typeManager = new TypeRegistry();
-        FunctionRegistry functionRegistry = new FunctionRegistry(typeManager, new BlockEncodingManager(typeManager), new FeaturesConfig());
-
         Expression expression = new LongLiteral("1");
         ImmutableMap.Builder<NodeRef<Expression>, Type> types = ImmutableMap.builder();
         types.put(NodeRef.of(expression), BIGINT);
         for (int i = 0; i < 100; i++) {
             expression = new CoalesceExpression(expression, new LongLiteral("2"));
             types.put(NodeRef.of(expression), BIGINT);
         }
-        SqlToRowExpressionTranslator.translate(expression, SCALAR, types.build(), functionRegistry, typeManager, TEST_SESSION, true);
+        translateAndOptimize(expression, types.build());
+    }
+
+    private RowExpression translateAndOptimize(Expression expression)
+    {
+        return translateAndOptimize(expression, getExpressionTypes(expression));
+    }
+
+    private RowExpression translateAndOptimize(Expression expression, Map<NodeRef<Expression>, Type> types)
+    {
+        return SqlToRowExpressionTranslator.translate(expression, SCALAR, types, metadata.getFunctionRegistry(), metadata.getTypeManager(), TEST_SESSION, true);
+    }
+
+    private Expression simplifyExpression(Expression expression)
```

Comment:
it is not yet used

Reply:
i am aware. This makes the second commit easier to read. Do you think it's bad?